### ISSUE TITLE
RFC Performance Improvement Opportunity

### ISSUE THREAD

**Issue by gdborton** [webpack/enhanced-resolve#88]
**TLDR:** Removing `createInnerCallback` sees a 2x performance boost.

I've been doing a lot of profiling of our webpack builds to see if there are any savings to be had. While doing this I noticed that there is a ton of garbage collection happening, and through that I found my way to this repository.

It seems that `createInnerCallback` creates many large arrays (full of strings) that go unused unless you're passing the `--verbose` flag, and your build has errors.  There may be other use cases, but that's all I've been able to find.

How can we limit the impact here?  I'd love to remove this log entirely as I don't see the value of it as an end user of webpack (webpack 3? 🤞 ), especially since disabling nets a huge performance win.  If that's not possible can we expose an option to disable?

## Charts!

 - It's a little bit difficult to highlight, but notice in this call stack that after each `loggingCallbackWrapper` there is a garbage collection event.

![image](https://user-images.githubusercontent.com/4172067/26909158-deac7ecc-4bb3-11e7-9b0e-45ef725f1816.png)

 - Here the calls to `createInnerCallback` have been removed (no more garbage collection).

![image](https://user-images.githubusercontent.com/4172067/26909186-fc93fb54-4bb3-11e7-8d40-87ecd71d1e2a.png)

Bottom up profile of our build before any changes:
![image](https://user-images.githubusercontent.com/4172067/26908759-a2302eaa-4bb1-11e7-84b5-56b1ea5869b2.png)

After removing all the `createInnerCallback` calls:
![image](https://user-images.githubusercontent.com/4172067/26908808-e5c7731c-4bb1-11e7-9885-57594e7997e5.png)

Removing `createInnerCallback` calls + running with `--max-semi-space-size=1000`:
![image](https://user-images.githubusercontent.com/4172067/26908914-7e7220da-4bb2-11e7-831b-05fa9ca227df.png)

**Comment by gdborton** [comment]
@TheLarkInn @sokra thoughts?

**Comment by TheLarkInn** [comment]
Wow. This def seems like a huge win. Thanks for tagging us (GitHub notifications are the worst :( )

I'm all for this change. do you have a partial implementation so far?

**Comment by TheLarkInn** [comment]
@sokra we could probably decouple the logger a bit better and just pass a debug flag to the resolver/factory. 

TBH I only ever see this information actually show when I'm working on enhanced-resolve locally, and never during a webpack build even with `--verbose`. So maybe we could remove it entirely except for when local development like you mentioned.

**Comment by gdborton** [comment]
I don't have any implementation in mind as I wasn't sure how else the log might get used. Personally I'd rather see it removed completely (or maybe only available through an internal flag?).

I don't really like the idea of adding an option for disabling a feature no one uses. Especially considering the biggest hurdle to adopting webpack is the config.

I don't mind doing the work though if we can find a reasonable solution.

**Comment by TheLarkInn** [comment]
I agree I think we should just have it as part of our test fixtures and local Dev if needed. It is very helpful for resolver plugin Dev also.

**Comment by niieani** [comment]
If you run `webpack` programatically, you can use this workaround (prior to requiring `webpack`) to get the above mentioned speed bump:

```js
// useful Webpack performance benefit until this is resolved: https://github.com/webpack/enhanced-resolve/issues/88
const Module = require('module')

function createCallbackMock(callback, options, message, messageOptional) {
  return function callbackMock() {
    // eslint-disable-next-line prefer-rest-params
    return callback.apply(this, arguments)
  }
}

// magic for mocking modules
const originalLoad = Module._load
function mockModuleLoader(...mocks) {
  Module._load = function load(request, parent, isMain) {
    for (const mock of mocks) {
      const mockedModuleExports = mock(request, parent, isMain)
      if (mockedModuleExports !== undefined) {
        return mockedModuleExports
      }
    }
    return originalLoad.apply(undefined, arguments)
  }
}

// let's mock the 'createInnerCallback' file:
const createCallbackRegExp = /\/enhanced-resolve\/lib\/createInnerCallback.js$/
mockModuleLoader(
  (request, parent, isMain) => {
    const filename = Module._resolveFilename(request, parent, isMain)
    return createCallbackRegExp.test(filename) ? createCallbackMock : undefined
  }
)
```

This effectively replaces the `module.exports` of the `createInnerCallback.js` file with a dummy method.

**Comment by TheLarkInn** [comment]
Actually now that I think about this, we could leave as a movie empty fn and patch the logger when needed.

**Comment by sokra** [comment]
we can resolve without log first and in case of an error we do an additional resolve with log.

**Comment by ljharb** [comment]
I like that approach in general; it pushes all the perf cost to the error path, which doesn't matter if it's fast or not.

**Comment by mikesherov** [comment]
OK, so it seems like we make a new function like `doResolveWithLogging` that'll call `doResolve` and on error, recall `doResolve` with the createInnerCallback version attached?

**Comment by TheLarkInn** [comment]
That sounds right to me.
